docs/design/marketing-offer-program.md

```diff
@@ -67,54 +67,146 @@ The credit is automatically applied to the account and will have a max limit tha
+    status - enum[ACTIVE, DEFAULT, DONE]
     PRIMARY KEY (id)
 ```
 
 **user_credit table**
 ```sql
+    id - int
     user_id - bytea
     offer_id - int
-    credits_earned - int
-    credits_used - int
+    credits_earned_in_cents - int
+    credits_used_in_cents - int
     credit_type - enum[AWARD, INVITEE, NO_TYPE]
     expires_at - timestamp
     created_at - timestamp
     referred_by - bytea (nullable)
     FOREIGN KEY (offer_id)
     FOREIGN KEY (referred_by)
-    PRIMARY KEY (user_id)
+    FOREIGN KEY (user_id)
+    PRIMARY KEY (id)
 ```
 
 **user table**
 ```sql
     total_Referred - int
 ```
 
-### Offer Program Service
+### Marketing Service
 
-**satellite/marketing/offers.go**
-- Create offers interface to interact with offer table
+**satellite/marketing/service.go**
+```golang
+func (m *marketing) GetCurrentOffer(ctx context.Context, offerStatus OfferStatus) (*Offer, error) {
+  offer, err := m.db.Marketing().Offers().Get_Offer_By_Status(ctx, offerStatus)
+  if err == sql.ErrNoRows && offerStatus == Active {
+    offer, err = m.db.Marketing().Offers().Get_Offer_By_Status(ctx, Default)
+    if err != nil {
+      return nil, Error.Wrap(err)
+    }
+  }
+  if err != nil {
+    return nil, Error.Wrap(err)
+  }
+
+  return offer, nil
+}
+
+func (m *marketing) StopOffer(ctx context.Context, offerId Offer.ID) error {
+  o := UpdateOffer{
+    Status: Done,
+    ExpiresAt: time.Now(),
+  }
+  err := m.db.Marketing().Offers().Update_Offer_By_ID(ctx, offerId, o)
+  if err != nil {
+    return Error.Wrap(err)
+  }
+
+  return nil
+}
+
+func (m *marketing) Create(ctx context.Context, offer Offer) error {
+  if offer.Status == Default {
+    offer.ExpiresAt = time.Now().AddDate(100, 0, 0)
+  }
+
+  err := m.db.Marketing().Offers().Create_Offer(ctx, offer)
+  if err != nil {
+    return Error.Wrap(err)
+  }
+
+  return nil
+}
 
+func (m *marketing) ListAllOffers(ctx context.Context) ([]Offers, error)
+```
+
+**satellite/marketing/offers.go**
 ```golang
-type DB interface {
+type Offers interface {
   ListAllOffers(ctx context.Context) ([]Offer, error)
-  GetOfferById(ctx context.Context, offerId Offer.ID) (Offer, error)
-  Update(ctx context.Context, offer *Offer) (*Offer, error)
-  Delete(ctx context.Context, offerId Offer.ID)
+  GetCurrentOffer(ctx context.Context, offerId Offer.ID) (Offer, error)
   Create(ctx context.Context, offer *Offer)
 }
 
-func (offer *Offer) IsUpdatable() bool {
-  return offer.Status == NO_STATUS
+type OfferStatus int
+const (
+  Done OfferStatus = 0
+  Active OfferStatus = 1
+  Default OfferStatus = 2
+)
+
+type UpdateOffer struct {
+  Status OfferStatus
+  ExpiresAt time.time
+}
+```
+
+**satellite/satellitedb/offers.go**
+```golang
+func (o *offersDB) Create(ctx context.Context, offer *marketing.Offer) error {
+  tx, err := o.db.Open(ctx)
+  if err != nil {
+    return marketing.OfferError.Wrap(err)
+  }
+
+  _, err := tx.Get_Offer_By_Status(ctx, dbx.Offer_Status(offer.Status))
```

Comment:
I am not sure I understand how you can `Get_Offer_By_Status`, `Update_Offer_By_Status`, etc... when the status is a simple enum and not a key. How does it know which offer to update or get (less applicable for create)

Reply:
That's a good point. I think the idea is that there will only be one active offer at a time. But maybe to be more secure, I should get the offer that has an active status with its expiration date to be greater than `NOW()` and the latest `created_at` ? What do you think?

Reply:
Oh, I didn't realize there would only be one active offer at a time. If this is true, then I don't actually see a problem with doing things this way. I should review the unchanged parts of the doc to get a better understanding

Reply:
That's a business requirement, but it's definitely a good point to make sure that we only get one offer back from the database as current offer

Reply:
Wait so there will never be two offers with different IDs associated with different users that are both active? I still think I could be misunderstanding

Reply:
No worries, it's complicated for sure. Ideally, there will always be two ongoing offers. One( status: active) is used for referral program, one(status: default) is for new user signup. If the referral offer is stopped or expired, it will use the default offer until a new offer starts